File: src/main/java/cn/labzen/web/ap/LabzenWebProcessor.java
package cn.labzen.web.ap;

import cn.labzen.web.ap.config.Config;
import cn.labzen.web.ap.config.ConfigLoader;
import cn.labzen.web.ap.internal.context.AnnotationProcessorContext;
import cn.labzen.web.ap.internal.context.ControllerContext;
import cn.labzen.web.ap.processor.InternalProcessor;
import com.google.auto.service.AutoService;
import com.google.common.collect.Sets;

import javax.annotation.processing.*;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.TypeElement;
import javax.lang.model.util.ElementKindVisitor14;
import javax.tools.Diagnostic;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.Comparator;
import java.util.List;
import java.util.ServiceLoader;
import java.util.Set;

@SupportedSourceVersion(SourceVersion.RELEASE_21)
@SupportedAnnotationTypes("cn.labzen.web.annotation.LabzenController")
@AutoService(Processor.class) // 真邪了门儿了，最朴实的方式javax.annotation.processing.Processor文件，编译会报错，用这个注解自动生成SPI文件就没事儿，MLGB！
public class LabzenWebProcessor extends AbstractProcessor {

  private static final Comparator<InternalProcessor> PROCESSOR_COMPARATOR = Comparator.comparing(InternalProcessor::priority);

  private final Set<TypeElement> processedControllers = Sets.newConcurrentHashSet();
  private final Set<DeferredController> deferredControllers = Sets.newConcurrentHashSet();

  private AnnotationProcessorContext annotationProcessorContext;

  @Override
  public synchronized void init(ProcessingEnvironment processingEnv) {
    super.init(processingEnv);

    Config config = ConfigLoader.load(processingEnv.getFiler());

    this.annotationProcessorContext = new AnnotationProcessorContext(processingEnv.getElementUtils(), processingEnv.getTypeUtils(), processingEnv.getMessager(), processingEnv.getFiler(), config);
  }

  @Override
  public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
    if (!roundEnv.processingOver()) {
      List<ControllerContext> deferredControllerContexts = getAndResetDeferredControllers();
      deferredControllerContexts.forEach(this::processEachController);

      List<ControllerContext> controllerContexts = getControllers(annotations, roundEnv);
      controllerContexts.forEach(this::processEachController);
    } else if (!deferredControllers.isEmpty()) {
      outputFailedControllers();
    }

    return false;
  }

  private void outputFailedControllers() {
    deferredControllers.forEach(deferred -> {
      TypeElement deferredElement = annotationProcessorContext.elements().getTypeElement(deferred.element().getQualifiedName());
      annotationProcessorContext.messaging().warning("LabzenWebProcessor: 无法实现 Controller " + deferredElement.getQualifiedName());
    });
  }

  private List<ControllerContext> getAndResetDeferredControllers() {
    return deferredControllers.stream().map(controller -> new ControllerContext(controller.element(), annotationProcessorContext)).toList();
  }

  private List<ControllerContext> getControllers(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
    return annotations.stream()
      .filter(annotation -> annotation.getKind() == ElementKind.ANNOTATION_TYPE)
      .flatMap(annotation ->
        roundEnv.getElementsAnnotatedWith(annotation).stream()
          .map(this::asTypeElement)
          .map(element -> new ControllerContext(element, annotationProcessorContext))
      ).toList();
  }

  private void processEachController(ControllerContext context) {
    if (processedControllers.contains(context.getSource())) {
      return;
    }

    List<InternalProcessor> processors = getProcessors();
    try {
      for (InternalProcessor processor : processors) {
        processor.process(context);
      }

      processedControllers.add(context.getSource());
    } catch (Throwable e) {
      deferredControllers.add(new DeferredController(context.getSource()));
      handleUncaughtError(context.getSource(), e);
    }
  }

  private void handleUncaughtError(Element element, Throwable e) {
    StringWriter sw = new StringWriter();
    e.printStackTrace(new PrintWriter(sw));

    String reportableStacktrace = sw.toString().replace(System.lineSeparator(), "  ");

    processingEnv.getMessager().printMessage(Diagnostic.Kind.ERROR, "Internal error in the mapping processor: " + reportableStacktrace, element);
  }

  private List<InternalProcessor> getProcessors() {
    return ServiceLoader.load(InternalProcessor.class, this.getClass().getClassLoader()).stream().map(ServiceLoader.Provider::get).sorted(PROCESSOR_COMPARATOR).toList();
  }

  private TypeElement asTypeElement(Element element) {
    return element.accept(new ElementKindVisitor14<TypeElement, Void>() {
      @Override
      public TypeElement visitTypeAsClass(TypeElement e, Void unused) {
        return e;
      }

      @Override
      public TypeElement visitTypeAsInterface(TypeElement e, Void unused) {
        return e;
      }
    }, null);
  }
}

File: src/main/kotlin/cn/labzen/web/ap/LabzenWebProcessor.kt
package cn.labzen.web.ap

import cn.labzen.web.ap.config.ConfigLoader
import cn.labzen.web.ap.exception.TypeNotAvailableException
import cn.labzen.web.ap.internal.context.AnnotationProcessorContext
import cn.labzen.web.ap.internal.context.ControllerContext
import cn.labzen.web.ap.processor.InternalProcessor
import java.io.PrintWriter
import java.io.StringWriter
import java.util.*
import javax.annotation.processing.AbstractProcessor
import javax.annotation.processing.ProcessingEnvironment
import javax.annotation.processing.RoundEnvironment
import javax.annotation.processing.SupportedAnnotationTypes
import javax.lang.model.SourceVersion
import javax.lang.model.element.Element
import javax.lang.model.element.ElementKind
import javax.lang.model.element.TypeElement
import javax.lang.model.util.ElementKindVisitor9
import javax.tools.Diagnostic

@SupportedAnnotationTypes("cn.labzen.web.annotation.LabzenController")
class LabzenWebProcessor : AbstractProcessor() {

  private lateinit var annotationProcessorContext: AnnotationProcessorContext
  private val processedControllers = mutableSetOf<TypeElement>()
  private val deferredControllers = mutableSetOf<DeferredController>()

  override fun init(processingEnv: ProcessingEnvironment) {
    super.init(processingEnv)

    val config = ConfigLoader.load(processingEnv.filer)

    annotationProcessorContext =
      AnnotationProcessorContext(
        processingEnv.elementUtils,
        processingEnv.typeUtils,
        processingEnv.messager,
        processingEnv.filer,
        config,
      )
  }

  override fun getSupportedSourceVersion(): SourceVersion =
    SourceVersion.latestSupported()

  override fun process(
    annotations: Set<TypeElement>,
    roundEnv: RoundEnvironment
  ): Boolean {
    if (!roundEnv.processingOver()) {
      val deferredControllerContexts = getAndResetDeferredControllers()
      deferredControllerContexts.forEach(::processEachController)

      val controllerContexts = getControllers(annotations, roundEnv)
      controllerContexts.forEach(::processEachController)
    } else if (deferredControllers.isNotEmpty()) {
      outputFailedControllers()
    }

    return false
  }

  private fun outputFailedControllers() {
    deferredControllers.forEach { dc ->
      val deferredElement = annotationProcessorContext.elementUtils.getTypeElement(dc.element.qualifiedName)
      annotationProcessorContext.messages.warning("LabzenWebProcessor: 无法实现 Controller ${deferredElement.qualifiedName}, 因为无法引用类 ${dc.referencedClass}")
    }
  }

  private fun getAndResetDeferredControllers(): List<ControllerContext> {
    val readied = deferredControllers.filter {
      try {
        Class.forName(it.referencedClass)
        true
      } catch (e: Exception) {
        false
      }
    }
    val readiedContexts = readied.map {
      ControllerContext(it.element, annotationProcessorContext)
    }

    deferredControllers.removeAll(readied.toSet())

    return readiedContexts
  }

  private fun getControllers(annotations: Set<TypeElement>, roundEnv: RoundEnvironment): List<ControllerContext> {
    return annotations.filter {
      it.kind == ElementKind.ANNOTATION_TYPE
    }.flatMap { annotation ->
      val annotatedElements = roundEnv.getElementsAnnotatedWith(annotation).map(::asTypeElement)
      annotatedElements.map { element ->
        ControllerContext(element, annotationProcessorContext)
      }
    }
  }

  private fun processEachController(context: ControllerContext) {
    if (processedControllers.contains(context.source)) {
      println("%%%%%%%%%%%%%%  ${context.source.simpleName}")
      return
    }

    val processors = getProcessor()
    try {
      for (processor in processors) {
        processor.process(context)
      }

      processedControllers.add(context.source)
    } catch (e: TypeNotAvailableException) {
      deferredControllers.add(DeferredController(context.source, e.fqcn))
    } catch (e: Throwable) {
      handleUncaughtError(context.source, e)
    }
  }

  private fun handleUncaughtError(element: Element, thrown: Throwable) {
    val sw = StringWriter()
    thrown.printStackTrace(PrintWriter(sw))

    val reportableStacktrace = sw.toString().replace(System.lineSeparator(), "  ")

    processingEnv.messager.printMessage(
      Diagnostic.Kind.ERROR, "Internal error in the mapping processor: $reportableStacktrace", element
    )
  }

  private fun getProcessor(): List<InternalProcessor> =
    ServiceLoader.load(InternalProcessor::class.java, this.javaClass.classLoader)
      .sortedWith(processorComparator)
      .toList()

  private fun asTypeElement(element: Element): TypeElement =
    element.accept(
      object : ElementKindVisitor9<TypeElement, Void?>() {
        override fun visitTypeAsInterface(e: TypeElement, p: Void?): TypeElement {
          return e
        }

        override fun visitTypeAsClass(e: TypeElement, p: Void?): TypeElement {
          return e
        }
      }, null
    )

  data class DeferredController(val element: TypeElement, val referencedClass: String)

  companion object {
    private val processorComparator: Comparator<InternalProcessor> = Comparator.comparing(
      InternalProcessor::priority
    )
  }
}